torch/distributed/_composable/fsdp/_fsdp_clip_grad_norm.py

```diff
@@ -0,0 +1,117 @@
+import functools
+import math
+import warnings
+from typing import cast, Iterable, List, Union
+
+import torch
+import torch.distributed as dist
+from torch.distributed._tensor import DTensor
+
+
+@torch.no_grad()
+def _get_grad_norm(
+    local_grads: List[torch.Tensor],
+    norm_type: float,
+    default_device: torch.device,
+) -> torch.Tensor:
+    """
+    Returns the gradient norm of ``local_grads``, where the gradients are
+    viewed as a single vector.
+    """
+    if len(local_grads) == 0:
+        return torch.tensor(0.0, dtype=torch.float32, device=default_device)
+    vector_norm_fn = functools.partial(
+        torch.linalg.vector_norm, ord=norm_type, dtype=torch.float32
+    )
+    grad_norm = vector_norm_fn(
+        torch.stack([vector_norm_fn(grad) for grad in local_grads])
+    )
+    return grad_norm
+
+
+@torch.no_grad()
+def clip_grad_norm_(
+    parameters: Union[DTensor, Iterable[DTensor]],
+    max_norm: float,
+    norm_type: float = 2.0,
+) -> torch.Tensor:
+    """
+    Clip the gradient norm of all parameters in-place. The norm is computed
+    over all gradients together, as if they were concatenated into a single
+    vector, including communication for sharded gradients.
+
+    Args:
+        max_norm (float): Max norm of the gradients.
+        norm_type (float): Type of the used p-norm. Can be ``float('inf')`` for
+            infinity norm.
+
+    Returns:
+        torch.Tensor: Total norm of the gradients in ``torch.float32``.
+
+    .. note:: If the passed in ``parameters`` is empty, then this returns the
+        zero tensor on CPU, following the convention of
+        :func:`torch.nn.utils.clip_grad_norm_`. Otherwise, this returns the
+        total norm on the same device as the parameters' gradients.
+
+    .. warning:: This needs to be called on all ranks since it uses collective
+        communications.
+    """
+    if isinstance(parameters, torch.Tensor):
+        parameters = [cast(DTensor, parameters)]
+    else:
+        parameters = list(parameters)
+    if len(parameters) == 0:
+        # Use default device following `torch.nn.utils.clip_grad_norm_()` but
+        # use fp32 for consistency with the non-empty case
+        zero_norm = torch.tensor(0.0, dtype=torch.float32)
+        warnings.warn(
+            f"Called clip_grad_norm_() on rank {dist.get_rank()} with empty "
+            f"parameters -- returning zero norm on CPU in fp32"
+        )
+        return zero_norm
+
+    for i, param in enumerate(parameters):
```

Comment:
The current implementation of `clip_grad_norm_()` requires `DTensor` inputs but does not go through `DTensor` op dispatch.

Would it be possible to implement `clip_grad_norm_()` through `DTensor` op dispatch? Otherwise, handling all different sharding placements will be complicated.